pennylane/transforms/decompositions/clifford_plus_t/cliffordt_transform.py

```diff
@@ -0,0 +1,272 @@
+import pennylane as qml
+from pennylane.transforms.core import transform
+from pennylane.tape import QuantumTape
+from pennylane.transforms.optimization import (
+    cancel_inverses,
+    commute_controlled,
+    merge_rotations,
+    remove_barrier,
+)
+
+# Single qubits Clifford+T gates in PL
+_CLIFFORD_T_ONE_GATES = [
+    qml.Identity,
+    qml.PauliX,
+    qml.PauliY,
+    qml.PauliZ,
+    qml.Hadamard,
+    qml.S,
+    qml.SX,
+    qml.T,
+]
+
+# Two qubits Clifford+T gates in PL
+_CLIFFORD_T_TWO_GATES = [
+    qml.CNOT,
+    qml.CY,
+    qml.CZ,
+    qml.SWAP,
+    qml.ISWAP,
+]
+
+_PARAMETER_GATES = [qml.RX, qml.RY, qml.RZ, qml.Rot, qml.GlobalPhase]
+
+_CLIFFORD_T_GATES = _CLIFFORD_T_ONE_GATES + _CLIFFORD_T_TWO_GATES
+
+
+# pylint: disable= bare-except, unnecessary-lambda-assignment
+def check_clifford_op(op):
+    r"""Check if an operator is Clifford or not.
+
+    For a given unitary operator :math:`U` acting on :math:`N` qubits, this method checks that the
+    transformation :math:`UPU^{\dagger}` maps the Pauli tensor products :math:`P = {I, X, Y, Z}^{\otimes N}`
+    to Pauli tensor products with O(N * 8^N) time complexity when using naive matrix multiplication.
+
+    Args:
+        op: the operator that needs to be tested
+
+    Returns:
+        Bool that represents whether the provided operator is Clifford or not.
+    """
+
+    num_qubits = len(op.wires)
+    try:
+        op_matrix = qml.matrix(op, wire_order=range(num_qubits))
+    except:
+        return False
+
+    pauli_terms = qml.pauli_decompose(op_matrix, check_hermitian=False)
+    pauli_group = lambda x: [qml.Identity(x), qml.PauliX(x), qml.PauliY(x), qml.PauliZ(x)]
+
+    pauli_coves = []
+    try:
+        pauli_qubit = [
+            qml.prod(*pauli) for pauli in product(*(pauli_group(idx) for idx in range(num_qubits)))
+        ]
+    except:
+        pauli_qubit = [qml.Identity(0)]
+    pauli_qubit = [
+        qml.pauli.pauli_sentence(op).hamiltonian(wire_order=range(num_qubits)) for op in pauli_qubit
+    ]
+
+    for prod in product([pauli_terms], pauli_qubit, [pauli_terms]):
+        # hopefully op_math.prod scales better than matrix multiplication, i.e., O((2^N)^3)
+        upu = qml.pauli.pauli_sentence(qml.prod(*prod))
+        upu.simplify()
+        upu2 = upu.hamiltonian(wire_order=range(num_qubits))
+        if len(upu2.ops) == 1:
+            if not isinstance(upu2.ops[0], qml.Identity):
+                pauli_coves.append(any((qml.equal(upu2.ops[0], tm) for tm in pauli_qubit)))
+        else:
+            pauli_coves.append(False)
+
+    return all(pauli_coves)
+
+
+def _check_clifford_t(op):
+    """Checks whether the gate is in the standard Clifford+T basis"""
+    # Save time and check from the pre-computed list
+    if any(
+        (
+            isinstance(op, gate) or isinstance(getattr(op, "base", None), gate)
+            for gate in _CLIFFORD_T_GATES
+        )
+    ):
+        return True
+    return check_clifford_op(op)
```

Comment:
Yep. Also, I don't think it would be heavy since most of the gates that we have are not more than three qubits, which should be rather quick to check.

Reply:
:thinking: But we will still have common multi-qubit ops like QFT that can involve a lot of qubits and would be expensive to evaluate the matrix for.

I think we can have a `lazy=True` arg that returns `False` here.

Reply:
I am pretty sure they won't be able to reach this check without getting expanded to smaller gates which shouldn't be that heavy.